### PR TITLE
develop to master

### PR DESCRIPTION
**Pull request by budgetpreneur**

### CODE REVIEW

File: .pubnub.yml [hunk 1]
```diff
@@ -1,8 +1,19 @@
 name: c-sharp
-version: "4.1.0.0"
+version: "4.1.1.0"
```
Comment (davidnub) [inline .pubnub.yml]: Shouldn't this be `4.2.0.0` if we're adding message actions?

File: src/Api/PubnubApi/Builder/UriUtil.cs [hunk 2]
```diff
@@ -103,7 +103,7 @@ public static string EncodeUriComponent(bool forPamSign, string s, PNOperationTy
             }
 
             encodedUri = o.ToString();
-            if (type == PNOperationType.PNHereNowOperation || type == PNOperationType.PNHistoryOperation || type == PNOperationType.Leave || type == PNOperationType.PNHeartbeatOperation || type == PNOperationType.PushRegister || type == PNOperationType.PushRemove || type == PNOperationType.PushGet || type == PNOperationType.PushUnregister)
+            if (type == PNOperationType.PNHereNowOperation || type == PNOperationType.PNHistoryOperation || type == PNOperationType.PNFetchHistoryOperation || type == PNOperationType.Leave || type == PNOperationType.PNHeartbeatOperation || type == PNOperationType.PushRegister || type == PNOperationType.PushRemove || type == PNOperationType.PushGet || type == PNOperationType.PushUnregister)
```
Comment (davidnub) [inline src/Api/PubnubApi/Builder/UriUtil.cs]: This is getting kind of crazy, we should refactor it some other time.

File: .pubnub.yml (hunk 1 above)
Comment (budgetpreneur) [reply .pubnub.yml]: ok..will change it to 4.2.0.0

File: src/Api/PubnubApi/Builder/UriUtil.cs (hunk 2 above)
Comment (budgetpreneur) [reply src/Api/PubnubApi/Builder/UriUtil.cs]: will refactor for this version for review

Comment (budgetpreneur) [reply src/Api/PubnubApi/Builder/UriUtil.cs]: code refactored. please review.

File: .pubnub.yml (hunk 1 above)
Comment (budgetpreneur) [reply .pubnub.yml]: done

File: src/Examples/PubnubApi.ConsoleExample/packages.config [hunk 3]
```diff
@@ -1,12 +1,8 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="AsyncBridge.JetBrains" version="0.1.1.0" targetFramework="net40" />
-  <package id="Microsoft.Bcl" version="1.1.8" targetFramework="net461" />
-  <package id="Microsoft.Bcl.Async" version="1.0.168" targetFramework="net461" />
-  <package id="Microsoft.Bcl.Build" version="1.0.14" targetFramework="net461" />
   <package id="Newtonsoft.Json" version="9.0.1" targetFramework="net461" />
   <package id="Portable.BouncyCastle" version="1.8.1.2" targetFramework="net461" />
-  <package id="Pubnub" version="4.1.0" targetFramework="net461" />
+  <package id="Pubnub" version="4.1.1" targetFramework="net461" />
```
Comment (davidnub) [inline src/Examples/PubnubApi.ConsoleExample/packages.config]: I think this needs to be `4.2.0` as well right?

File: src/Examples/PubnubApi.WinFormExample/PubnubApi.WinFormExample.csproj [hunk 4]
```diff
@@ -50,11 +49,10 @@
       <Private>True</Private>
     </Reference>
     <Reference Include="Newtonsoft.Json, Version=9.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Newtonsoft.Json.9.0.1\lib\net45\Newtonsoft.Json.dll</HintPath>
-      <Private>True</Private>
+      <HintPath>packages\Newtonsoft.Json.9.0.1\lib\net45\Newtonsoft.Json.dll</HintPath>
     </Reference>
-    <Reference Include="Pubnub, Version=4.0.16.0, Culture=neutral, PublicKeyToken=dc66f52ce6619f44, processorArchitecture=MSIL">
-      <HintPath>packages\Pubnub.4.0.17\lib\net45\Pubnub.dll</HintPath>
+    <Reference Include="Pubnub, Version=4.1.1.0, Culture=neutral, PublicKeyToken=dc66f52ce6619f44, processorArchitecture=MSIL">
```
Comment (davidnub) [inline src/Examples/PubnubApi.WinFormExample/PubnubApi.WinFormExample.csproj]: Should it be `4.2.0.0` ?

File: src/Examples/PubnubApi.WinFormExample/packages.config [hunk 5]
```diff
@@ -4,8 +4,8 @@
   <package id="Microsoft.Bcl.Async" version="1.0.168" targetFramework="net46" />
   <package id="Microsoft.Bcl.Build" version="1.0.14" targetFramework="net46" />
   <package id="Newtonsoft.Json" version="9.0.1" targetFramework="net46" />
-  <package id="Portable.BouncyCastle" version="1.8.1.1" targetFramework="net46" />
-  <package id="Pubnub" version="4.0.17" targetFramework="net46" />
+  <package id="Portable.BouncyCastle" version="1.8.1.2" targetFramework="net46" />
+  <package id="Pubnub" version="4.1.1" targetFramework="net46" />
```
Comment (davidnub) [inline src/Examples/PubnubApi.WinFormExample/packages.config]: `4.2.0` here

File: src/Examples/PubnubApi.ConsoleExample/packages.config (hunk 3 above)
Comment (budgetpreneur) [reply src/Examples/PubnubApi.ConsoleExample/packages.config]: updated.

File: src/Examples/PubnubApi.WinFormExample/PubnubApi.WinFormExample.csproj (hunk 4 above)
Comment (budgetpreneur) [reply src/Examples/PubnubApi.WinFormExample/PubnubApi.WinFormExample.csproj]: updated.

File: src/Examples/PubnubApi.WinFormExample/packages.config (hunk 5 above)
Comment (budgetpreneur) [reply src/Examples/PubnubApi.WinFormExample/packages.config]: updated.

**Pull request merged**